### PR TITLE
Remove much of the boilerplate around updating the XML UI in SetupChecker.

### PR DESCRIPTION
**Pull request by tomprince**

### CODE REVIEW

File: objects/SetupChecker/script.lua [hunk 1]
```diff
@@ -1295,3 +1221,84 @@ function tFind(table, needle)
...
+-- @param f The function that updates the element
+function matchRecurse(id, f)
+    return function (t)
+        if t.attributes.recurse ~= nil and string.match(t.attributes.recurse, id) then
+            if t.attributes.id == id then
+                f(t)
+            else
+                return true
+            end
+        end
```
Comment (iakona) [inline objects/SetupChecker/script.lua:1268]: so presumably there is an implied return nil for this function, but should probably be more explicit in nature

File: objects/SetupChecker/script.lua [hunk 2]
```diff
@@ -1295,3 +1221,84 @@ function tFind(table, needle)
...
+-- and return `true` if updateXml should recurse into the children of the element.
+-- Note: The functions should be prepared to be called on child elements, even if
+-- recursion was not requested.
+-- @param updateFunctions The list of functions to use to update the table
+function updateXml(updateFunctions)
+    local function recurse(t)
+        local shouldRecurse = false
+        for _, f in pairs(updateFunctions) do
+            if f(t) then
+                shouldRecurse = true
```
Comment (iakona) [inline objects/SetupChecker/script.lua:1238]: you should break after setting this to true

File: objects/SetupChecker/script.lua [hunk 3]
```diff
@@ -127,11 +127,11 @@ function onLoad(saved_data)
...
             -- queue up all dropdown changes as once
             Wait.frames(function()
-                local t = self.UI.getXmlTable()
-                t = updateAdversaryList(t)
-                t = updateScenarioList(t)
-                t = updateBoardLayouts(numPlayers, t)
-                self.UI.setXmlTable(t, {})
+                updateXml{
+                    updateAdversaryList(),
+                    updateBoardLayouts(numPlayers),
```
Comment (iakona) [inline objects/SetupChecker/script.lua:132]: is there any particular reason why you flipped scenario and board layout calls here?

File: objects/SetupChecker/script.lua (hunk 2 above)
Comment (tomprince) [reply objects/SetupChecker/script.lua:1238]: No. There may be another function that wants to mutate the table in the list of update functions.

Comment (iakona) [reply objects/SetupChecker/script.lua:1238]: oh shoot yea you're right

File: objects/SetupChecker/script.lua [hunk 4]
```diff
@@ -1295,3 +1221,85 @@ function tFind(table, needle)
...
+-- given function to the element with the given id. All the parent elements of the
+-- desired element must have the id present in the `recurse` attribute for this to
+-- find the element.
+-- @param id The id of the element to update
+-- @param f The function that updates the element
+function matchRecurse(id, f)
+    return function (t)
+        if t.attributes.recurse ~= nil and string.match(t.attributes.recurse, id) then
+            if t.attributes.id == id then
+                f(t)
```
Comment (iakona) [inline objects/SetupChecker/script.lua:1264]: assuming id field is unique (which it should be), then you should return f(t) right?

Comment (tomprince) [reply objects/SetupChecker/script.lua:1264]: I don't think so. The functions passed to `matchRecurse` all return `nil` and mutate `t`.

**Review by iakona — APPROVED** [review]

**Pull request merged**